### PR TITLE
UNR-561 view refactor

### PR DESCRIPTION
**Pull request by neprune**
#### Description
Make the `SpatialView` keep track of all hand-written components and provide a generic interface for retrieving them.
#### Tests
Manually tested by running the starter project and observing that it's functional.
#### Documentation
Internal refactor- no docs required.

### CODE REVIEW

File: SpatialGDK/Source/Private/Interop/SpatialView.cpp [hunk 1]
```diff
@@ -125,50 +125,63 @@ Worker_Authority USpatialView::GetAuthority(Worker_EntityId EntityId, Worker_Com
...
 {
-	EntityUnrealMetadataMap.Remove(Op.entity_id);
-	EntityACLMap.Remove(Op.entity_id);
+	EntityComponentMap.Remove(Op.entity_id);
 }
 
 void USpatialView::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
 {
-	if (Op.update.component_id == improbable::EntityAcl::ComponentId)
+	if (Op.update.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID)
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/Interop/SpatialView.cpp]: Why this change?

File: SpatialGDK/Source/Private/Interop/SpatialView.cpp [hunk 2]
```diff
@@ -125,50 +125,63 @@ Worker_Authority USpatialView::GetAuthority(Worker_EntityId EntityId, Worker_Com
...
-}
-
-improbable::EntityAcl* USpatialView::GetEntityACL(Worker_EntityId EntityId)
-{
-	if (TSharedPtr<improbable::EntityAcl>* EntityAclPtr = EntityACLMap.Find(EntityId))
-	{
-		return EntityAclPtr->Get();
+		if (TSharedPtr<improbable::ComponentStorageBase>* Component = ComponentStorageMap->Find(T::ComponentId))
+		{
+			return &static_cast<improbable::ComponentStorage<T>&>(*(Component->Get())).Get();
```
Comment (Vatyx) [inline SpatialGDK/Source/Private/Interop/SpatialView.cpp]: Dunno if this is completely right but `return &(static_cast<improbable::ComponentStorage<T>*>(Component->Get())->Get())`?

File: SpatialGDK/Source/Private/Interop/SpatialView.cpp (hunk 1 above)
Comment (neprune) [reply SpatialGDK/Source/Private/Interop/SpatialView.cpp]: Just for consistency with `OnAddComponent` which also uses `SpatialConstants`.

File: SpatialGDK/Source/Private/Interop/SpatialView.cpp [hunk 3]
```diff
@@ -125,50 +125,63 @@ Worker_Authority USpatialView::GetAuthority(Worker_EntityId EntityId, Worker_Com
...
 
 void USpatialView::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
 {
-	if (Op.update.component_id == improbable::EntityAcl::ComponentId)
+	if (Op.update.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID)
 	{
-		EntityACLMap[Op.entity_id]->ApplyComponentUpdate(Op.update);
+		if (improbable::EntityAcl* aclComponent = GetComponentData<improbable::EntityAcl>(Op.entity_id))
+		{
+			aclComponent->ApplyComponentUpdate(Op.update);
```
[the remark this answers is not in the document]
Comment (neprune) [reply SpatialGDK/Source/Private/Interop/SpatialView.cpp]: Yeah, weirdly `EntityACL` is the only component that actually implements an `ApplyComponentUpdate` at the moment. Will add implementations to the others. At this point I'll also define a lot of these methods as virtual in the base component struct.

[the remark this answers is not in the document]
Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialView.cpp]: Sounds reasonable.

[the remark this answers is not in the document]
Comment (neprune) [reply SpatialGDK/Source/Private/Interop/SpatialView.cpp]: Updated to actually manage updates for `Position` and `Rotation` - no point for other components AFAICS because we don't actually generate updates.

File: SpatialGDK/Source/Public/Schema/StandardLibrary.h [hunk 4]
```diff
@@ -15,7 +15,6 @@ using WriteAclMap = TMap<Worker_ComponentId, WorkerRequirementSet>;
 
 namespace improbable
 {
-
```
Comment (Vatyx) [inline SpatialGDK/Source/Public/Schema/StandardLibrary.h]: Super nit: put line back

File: SpatialGDK/Source/Public/Schema/Component.h [hunk 5]
```diff
@@ -12,4 +12,33 @@ struct Component
 	bool bIsDynamic = false;
 };
 
+class ComponentStorageBase
```
Comment (Vatyx) [inline SpatialGDK/Source/Public/Schema/Component.h]: What do you think about namespacing this with `improbable`?

File: SpatialGDK/Source/Public/EngineClasses/SpatialNetDriver.h [hunk 6]
```diff
@@ -119,7 +119,7 @@ class SPATIALGDK_API USpatialNetDriver : public UIpNetDriver
 	UPROPERTY()
 	UEntityRegistry* EntityRegistry;
 
-	TMap<Worker_EntityId_Key, USpatialActorChannel*> EntityToActorChannel;
+	TMap<Worker_EntityId, USpatialActorChannel*> EntityToActorChannel;
```
Comment (danielimprobable) [inline SpatialGDK/Source/Public/EngineClasses/SpatialNetDriver.h]: This will  not compile on linux, please revert back

File: SpatialGDK/Source/Public/Interop/SpatialView.h [hunk 7]
```diff
@@ -47,7 +47,6 @@ class SPATIALGDK_API USpatialView : public UObject
 	UPROPERTY()
 	USpatialSender* Sender;
 
-	TMap<Worker_EntityId_Key, TMap<Worker_ComponentId, Worker_Authority>> EntityComponentAuthorityMap;
-	TMap<Worker_EntityId_Key, TSharedPtr<improbable::UnrealMetadata>> EntityUnrealMetadataMap;
-	TMap<Worker_EntityId_Key, TSharedPtr<improbable::EntityAcl>> EntityACLMap;
+	TMap<Worker_EntityId, TMap<Worker_ComponentId, Worker_Authority>> EntityComponentAuthorityMap;
```
Comment (danielimprobable) [inline SpatialGDK/Source/Public/Interop/SpatialView.h]: These need to use `Worker_EntityId_Key` to compile on linux

File: SpatialGDK/Source/Public/Schema/Component.h (hunk 5 above)
Comment (danielimprobable) [reply SpatialGDK/Source/Public/Schema/Component.h]: it is in the improbable namespace already right?

File: SpatialGDK/Source/Public/SpatialConstants.h [hunk 8]
```diff
@@ -6,8 +6,6 @@
 
 #include <improbable/c_worker.h>
 
-using Worker_EntityId_Key = int64;
```
Comment (danielimprobable) [inline SpatialGDK/Source/Public/SpatialConstants.h]: We need this for linux, potentially add a comment for this so that we don't attempt this again

File: SpatialGDK/Source/Public/Utils/EntityRegistry.h [hunk 9]
```diff
@@ -52,6 +52,6 @@ class SPATIALGDK_API UEntityRegistry : public UObject
 
 	void RemoveFromRegistryImpl(const AActor* Actor, const Worker_EntityId& EntityId);
 
-	TMap<Worker_EntityId_Key, AActor*> EntityIdToActor;
+	TMap<Worker_EntityId, AActor*> EntityIdToActor;
```
Comment (danielimprobable) [inline SpatialGDK/Source/Public/Utils/EntityRegistry.h]: As the others, revert

File: SpatialGDK/Source/Private/Interop/SpatialView.cpp [hunk 10]
```diff
@@ -125,50 +125,79 @@ Worker_Authority USpatialView::GetAuthority(Worker_EntityId EntityId, Worker_Com
...
-}
-
-improbable::EntityAcl* USpatialView::GetEntityACL(Worker_EntityId EntityId)
-{
-	if (TSharedPtr<improbable::EntityAcl>* EntityAclPtr = EntityACLMap.Find(EntityId))
-	{
-		return EntityAclPtr->Get();
+		if (TUniquePtr<improbable::ComponentStorageBase>* Component = ComponentStorageMap->Find(T::ComponentId))
+		{
+			return &(static_cast<improbable::ComponentStorage<T>*>(Component->Get())->Get());
```
Comment (danielimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialView.cpp]: I'm pretty sure we should return weak pointers here instead of returning a raw pointer to a unique ptr. 

https://isocpp.org/blog/2013/07/quick-q-whats-the-weak-ptr-equivalent-for-unique-ptr-stackoverflow

File: SpatialGDK/Source/Public/SpatialConstants.h (hunk 8 above)
Comment (neprune) [reply SpatialGDK/Source/Public/SpatialConstants.h]: Could you explain why this breaks on Linux - just so I can add some context for the comment.

**Comment by neprune** [comment]
@danielimprobable Made the suggested changes, PTAL. I didn't move the `GlobalStateManager` 'tracking' logic into the `SpatialStaticComponentView` because I think it makes sense to keep the `NetDriver` as the owner of it.

File: SpatialGDK/Source/Private/Interop/SpatialDispatcher.cpp [hunk 11]
```diff
@@ -1,22 +1,22 @@
...
 
 DEFINE_LOG_CATEGORY(LogSpatialView);
 
-void USpatialView::Init(USpatialNetDriver* InNetDriver)
+void USpatialDispatcher::Init(USpatialNetDriver* InNetDriver)
 {
 	NetDriver = InNetDriver;
 	Receiver = InNetDriver->Receiver;
-	Sender = InNetDriver->Sender;
+    StaticComponentView = InNetDriver->StaticComponentView;
```
Comment (danielimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialDispatcher.cpp]: nit: spacing

File: SpatialGDK/Source/Private/Interop/SpatialDispatcher.cpp [hunk 12]
```diff
@@ -1,22 +1,22 @@
...
 
 DEFINE_LOG_CATEGORY(LogSpatialView);
 
-void USpatialView::Init(USpatialNetDriver* InNetDriver)
+void USpatialDispatcher::Init(USpatialNetDriver* InNetDriver)
 {
 	NetDriver = InNetDriver;
 	Receiver = InNetDriver->Receiver;
-	Sender = InNetDriver->Sender;
+        StaticComponentView = InNetDriver->StaticComponentView;
```
[the remark this answers is not in the document]
Comment (neprune) [reply SpatialGDK/Source/Private/Interop/SpatialDispatcher.cpp]: My bad, I made that change in vim which was using spaces. Fixed now.

File: SpatialGDK/Source/Public/Interop/SpatialDispatcher.h [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+private:
+	UPROPERTY()
+	USpatialNetDriver* NetDriver;
+
+	UPROPERTY()
+	USpatialReceiver* Receiver;
+
+	UPROPERTY()
+    USpatialStaticComponentView* StaticComponentView;
```
Comment (danielimprobable) [inline SpatialGDK/Source/Public/Interop/SpatialDispatcher.h]: Another spacing issue

Comment (neprune) [reply SpatialGDK/Source/Public/Interop/SpatialDispatcher.h]: Sorry, I'll take a closer look and make sure I've got them all

File: SpatialGDK/Source/Private/Interop/SpatialStaticComponentView.cpp [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+void USpatialStaticComponentView::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
+{
+	switch (Op.update.component_id)
+	{
+	case SpatialConstants::ENTITY_ACL_COMPONENT_ID:
+		if (improbable::EntityAcl* aclComponent = GetComponentData<improbable::EntityAcl>(Op.entity_id))
+		{
+			aclComponent->ApplyComponentUpdate(Op.update);
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialStaticComponentView.cpp]: What if we just grabbed the base component data, and add a virtual ApplyComponentUpdate function to the component class?

Comment (neprune) [reply SpatialGDK/Source/Private/Interop/SpatialStaticComponentView.cpp]: Yeah, I planned to do this for a load of the repeated component methods but didn't get around to it. Will at least do if for `ApplyComponentUpdate` before merging.

File: SpatialGDK/Source/Private/Interop/SpatialStaticComponentView.cpp [hunk 15]
```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialStaticComponentView.h"
+
+
```
Comment (danielimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialStaticComponentView.cpp]: Remove newline

File: SpatialGDK/Source/Public/SpatialConstants.h [hunk 16]
```diff
@@ -6,6 +6,10 @@
 
 #include <improbable/c_worker.h>
 
+
```
Comment (danielimprobable) [inline SpatialGDK/Source/Public/SpatialConstants.h]: Remove newline
